src/org/rascalmpl/parser/gtd/SGTDBF.java

```diff
@@ -781,78 +830,108 @@ private boolean findFirstStacksToReduce(){
 			DoubleStack<AbstractStackNode<P>, AbstractNode> terminalsTodo = todoLists[i];
 			if(!(terminalsTodo == null || terminalsTodo.isEmpty())){
 				stacksWithTerminalsToReduce = terminalsTodo;
-				
+
 				location += i;
-				
+
 				queueIndex = i;
-				
+
 				return true;
 			}
 		}
-		
+
 		if (recoverer != null) {
+			if (debugListener != null) {
```

Comment:
can we make an instance of `debugListener` that just drops everything (`NopDebugListener`)? so the `if` aren't needed everywhere?

Reply:
We definitely can. I think the original idea was that an `if` with a `null` check is faster than a no-op method call but that might not really be the case. Removing all the `if (debugListener != null)` checks definitely improved code quality.